### PR TITLE
masterdata translation

### PR DESCRIPTION
**Pull request by dageorgero**

### CODE REVIEW

File: services/masterdata/src/assets/i18n/en.json [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+			},
+			"masterdataTypeDetails": {
+				"header": "Masterdata-type Details",
+				"modelFields": {
+					"id": "Id",
+					"code": "Code",
+					"name": "Name",
+					"description": "Description",
+					"metadata": "Metadata",
+					"masterdatasCount": "Masterdatas Count"
```
Comment (Johanh-od) [inline services/masterdata/src/assets/i18n/en.json]: Shouldn't this be Master data item count?

File: services/masterdata/src/assets/i18n/nl.json [hunk 2]
```diff
@@ -0,0 +1,172 @@
+{
+	"masterdatamgmt": {
```
Comment (Johanh-od) [inline services/masterdata/src/assets/i18n/nl.json:8]: Could you use google translate or ask a native dutch speaker to translate this?

File: services/masterdata/src/lib/app-menu.ts [hunk 3]
```diff
@@ -2,9 +2,11 @@ import { MenuItem } from "@lens/app-abstract";
 
 export const menu: MenuItem[] = [
 	{
-		label: "Manage masterdata",
+		translationKey: "menu.title",
+		label: "Masterdata management",
```
Comment (Johanh-od) [inline services/masterdata/src/lib/app-menu.ts]: Label can be removed if translations are used

File: services/masterdata/src/lib/components/masterdata-type-details/masterdata-type-details.component.ts [hunk 4]
```diff
@@ -47,11 +49,9 @@ export class MasterdataTypeDetailsComponent implements OnInit {
...
 	prepareForDisplay(item: MasterdataType) {
-		return Object.entries(item).map(item => {
-			item[0] = item[0].toLowerCase();
-			if (item[0] === "masterdatascount") {
-				item[0] = `masterdatas ${item[0].replace("masterdatas", "")}`;
-			}
+		// "metadata" model field will have a special display!
+		return Object.entries(item).filter(item => item[0] !== "metadata").map(item => {
+			item[0] = this.translateService.instant(`masterdatamgmt.pages.masterdataTypeDetails.modelFields.${item[0]}`);
 			return item;
```
Comment (Johanh-od) [inline services/masterdata/src/lib/components/masterdata-type-details/masterdata-type-details.component.ts]: Unreachable code due to duplicate return statement.

File: services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html [hunk 5]
```diff
@@ -1,4 +1,11 @@
-<lens-panel header="{{ formTitle }} Masterdata-type '{{ item?.name }}'">
+<lens-panel
+	header="{{
```
Comment (Johanh-od) [inline services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html]: suggestion: move this kind of logic to typescript to keep the html clean

File: services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html [hunk 6]
```diff
@@ -131,7 +136,11 @@
 			<div class="form-actions">
 				<div>
 					<lens-button
-						label="{{ saveBtnText }}"
+						label="{{
```
Comment (Johanh-od) [inline services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html]: suggestion: move to typescript to keep html clean

File: services/masterdata/src/lib/components/masterdata-type-list/masterdata-type-list.component.ts [hunk 7]
```diff
@@ -14,7 +15,12 @@ export class MasterdataTypeListComponent implements OnInit {
...
 
-	constructor(private readonly service: MasterdataCrudHttpService, private readonly router: Router, private readonly activeRoute: ActivatedRoute) {}
+	constructor(
+		private readonly service: MasterdataCrudHttpService,
+		private readonly router: Router,
+		private readonly activeRoute: ActivatedRoute,
+		private readonly translateService: TranslateService
+	) {}
 
 	// eslint-disable-next-line @angular-eslint/no-empty-lifecycle-method
```
Comment (Johanh-od) [inline services/masterdata/src/lib/components/masterdata-type-list/masterdata-type-list.component.ts]: Comment can be removed. There is no empty lifecycle method anymore

File: services/masterdata/src/lib/components/masterdatas-details/masterdatas-details.component.ts [hunk 8]
```diff
@@ -50,11 +52,9 @@ export class MasterdatasDetailsComponent implements OnInit {
...
 	prepareForDisplay(item: Masterdata) {
-		return Object.entries(item).map(item => {
-			item[0] = item[0].toLowerCase().replace("masterdata", "");
-			if (["typeid", "typename"].includes(item[0])) {
-				item[0] = `type ${item[0].replace("type", "")}`;
-			}
+		// "metadata" model field will have a special display!
+		return Object.entries(item).filter(item => item[0] !== "metadata").map(item => {
+			item[0] = this.translateService.instant(`masterdatamgmt.pages.masterdataDetails.modelFields.${item[0]}`);
 			return item;
```
Comment (Johanh-od) [inline services/masterdata/src/lib/components/masterdatas-details/masterdatas-details.component.ts]: unreachable code due to duplicate return statement. Second time I see this code. Should this be moved to a separate file?

File: services/masterdata/src/lib/components/masterdatas-edit-form/masterdatas-edit-form.component.html [hunk 9]
```diff
@@ -1,5 +1,9 @@
 <lens-panel
-	header="{{ formTitle }} Masterdata to type '{{
+	header="{{
```
Comment (Johanh-od) [inline services/masterdata/src/lib/components/masterdatas-edit-form/masterdatas-edit-form.component.html]: suggestion: move to ts file (same for other occurrences in this file

File: services/masterdata/src/lib/components/masterdatas-list/masterdatas-list.component.ts [hunk 10]
```diff
@@ -23,7 +24,12 @@ export class MasterdatasListComponent implements OnInit {
...
 
-	constructor(private readonly service: MasterdataCrudHttpService, private readonly router: Router, private readonly activeRoute: ActivatedRoute) {}
+	constructor(
+		private readonly service: MasterdataCrudHttpService,
+		private readonly router: Router,
+		private readonly activeRoute: ActivatedRoute,
+		private readonly translateService: TranslateService
+	) {}
 
 	// eslint-disable-next-line @angular-eslint/no-empty-lifecycle-method
```
Comment (Johanh-od) [inline services/masterdata/src/lib/components/masterdatas-list/masterdatas-list.component.ts]: remove comment as this is obsolete. Same for line 40

**Review by Johanh-od — CHANGES_REQUESTED** [review]

File: services/masterdata/src/assets/i18n/en.json (hunk 1 above)
Comment (dageorgero) [reply services/masterdata/src/assets/i18n/en.json]: I prefer it like "masterdatas"

File: services/masterdata/src/lib/components/masterdata-type-details/masterdata-type-details.component.ts (hunk 4 above)
Comment (dageorgero) [reply services/masterdata/src/lib/components/masterdata-type-details/masterdata-type-details.component.ts]: @Johanh-od moved the duplicated code to a new service MasterdataHelperService

File: services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html (hunk 6 above)
Comment (dageorgero) [reply services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html:139]: @Johanh-od I need more details?

File: services/masterdata/src/lib/components/masterdatas-details/masterdatas-details.component.ts (hunk 8 above)
Comment (dageorgero) [reply services/masterdata/src/lib/components/masterdatas-details/masterdatas-details.component.ts]: @Johanh-od moved the duplicated code to a new service MasterdataHelperService

File: services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html (hunk 5 above)
Comment (dageorgero) [reply services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html]: Good point! I've updated the code.

File: services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html (hunk 6 above)
Comment (dageorgero) [reply services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.html]: Good point! I've updated the code.

File: services/masterdata/src/lib/components/masterdatas-edit-form/masterdatas-edit-form.component.html (hunk 9 above)
Comment (dageorgero) [reply services/masterdata/src/lib/components/masterdatas-edit-form/masterdatas-edit-form.component.html]: Good point! I've updated the code.

File: services/masterdata/src/assets/i18n/en.json (hunk 1 above)
Comment (jstorteboom) [reply services/masterdata/src/assets/i18n/en.json]: 'data', as in the english word, by nature is plural. so the 's' behind it doesn't make sense (which for other words is used to make them plural). So here it should be 'Masterdata count'.

File: services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.ts [hunk 11]
```diff
@@ -136,4 +140,15 @@ export class MasterdataTypeEditFormComponent implements OnInit {
 	navigateToListView() {
 		this.router.navigate(["/"]);
 	}
+
+	private buildTranslationTexts() {
+		this.languageService.onTranslationsLoaded(() => {
```
Comment (Johanh-od) [inline services/masterdata/src/lib/components/masterdata-type-edit-form/masterdata-type-edit-form.component.ts:146]: Tip: to make this code more compact you can also prepare the right resource key and then get if from the language service:


const mode = this.isAddForm ? "add" : "edit";
const headerKey = `masterdatamgmt.pages.masterdataTypeUpsert.${mode}FormTitle`
this.formHeaderText = this.translateService.instant(headerKey)

File: services/masterdata/src/lib/core/services/masterdata-helper.service.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
+import { Injectable } from "@angular/core";
+import { TranslateService } from "@ngx-translate/core";
+
+@Injectable()
+export class MasterdataHelperService {
```
Comment (Johanh-od) [inline services/masterdata/src/lib/core/services/masterdata-helper.service.ts]: Tip: using keywords like Helper is an indication of a class that don't have a high cohesion ([have too many responsibilities](https://medium.com/clarityhub/low-coupling-high-cohesion-3610e35ac4a6)). Its better to narrow it down to its specific function. E.g.: in this case you could use as class name MetadataRenderer, which covers the functionality a lot more than helper

File: services/masterdata/src/lib/core/services/masterdata-helper.service.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+	constructor(
+		private readonly translateService: TranslateService
+	) {}
+
+	prepareForDisplay<T extends object>(item: T, translationKey: string) {
+		// "metadata" model field will have a special display!
+		return Object.entries(item).filter(item => item[0] !== "metadata").map(item => {
+			item[0] = this.translateService.instant(`masterdatamgmt.pages.${translationKey}.modelFields.${item[0]}`);
+			return item;
```
Comment (Johanh-od) [inline services/masterdata/src/lib/core/services/masterdata-helper.service.ts]: Why is it needed to return item at line 14? You're also returning Object.entries on line 12. Seems unreachable code to me

File: services/masterdata/src/lib/core/services/masterdata-helper.service.ts (hunk 12 above)
Comment (dageorgero) [reply services/masterdata/src/lib/core/services/masterdata-helper.service.ts]: @Johanh-od MetadataRenderer or MasterdataRenderer ? So the full name of the service should be MetadataRendererService or MasterdataRendererService ?

File: services/masterdata/src/lib/core/services/masterdata-helper.service.ts (hunk 13 above)
Comment (dageorgero) [reply services/masterdata/src/lib/core/services/masterdata-helper.service.ts]: @Johanh-od good point, you are right!

File: services/masterdata/src/lib/core/services/masterdata-helper.service.ts (hunk 12 above)
Comment (Johanh-od) [reply services/masterdata/src/lib/core/services/masterdata-helper.service.ts]: Based on the comment at line 11 I guess its rendering metadata, so MetadataRenderer should be the name.

File: services/masterdata/src/lib/core/services/masterdata-renderer.service.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+export class MasterdataRendererService {
+	constructor(
+		private readonly translateService: TranslateService
+	) {}
+
+	prepareForDisplay<T extends object>(item: T, translationKey: string): Entries<T> {
+		// "metadata" model field will have a special display!
+		return (Object.entries(item) as Entries<T>).filter(item => item[0] !== "metadata").map(item => {
+			item[0] = this.translateService.instant(`masterdatamgmt.pages.${translationKey}.modelFields.${item[0].toString()}`);
+			return item;
```
Comment (Johanh-od) [inline services/masterdata/src/lib/core/services/masterdata-renderer.service.ts]: Why is the duplicate return statement still not removed?

**Review by Johanh-od — CHANGES_REQUESTED** [review]

File: services/masterdata/src/assets/i18n/nl.json [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+			},
+			"masterdataTypeDetails": {
+				"header": "Type stamgegevens Details",
+				"modelFields": {
+					"id": "Id",
+					"code": "Code",
+					"name": "Naam",
+					"description": "Beschrijving",
+					"metadata": "Metagegevens",
+					"masterdatasCount": "Stamgegevens tellen"
```
Comment (bart-vandervliet) [inline services/masterdata/src/assets/i18n/nl.json]: `count` is used as "number of masterdata", right? `tellen` is better translated as `aantal` then :).

File: services/masterdata/src/lib/core/services/masterdata-renderer.service.ts (hunk 14 above)
Comment (dageorgero) [reply services/masterdata/src/lib/core/services/masterdata-renderer.service.ts]: @Johanh-od fixed in commit "fixed issue with prepareForDisplay in MasterdataRendererService"

File: services/masterdata/src/lib/core/services/masterdata-helper.service.ts (hunk 12 above)
Comment (dageorgero) [reply services/masterdata/src/lib/core/services/masterdata-helper.service.ts]: @Johanh-od I've decided on MasterdataRendererService. I've also already another help method in this class: titleCase!

File: services/masterdata/src/lib/core/services/masterdata-renderer.service.ts (hunk 14 above)
Comment (jstorteboom) [reply services/masterdata/src/lib/core/services/masterdata-renderer.service.ts]: The ln 18 return is return in the map-callback function. The return on line 16 returns the result of the call to map. So this is correct.

Comment (Johanh-od) [reply services/masterdata/src/lib/core/services/masterdata-renderer.service.ts]: ah, now I see. Its a bit difficult to read due to the formatting

Comment (dageorgero) [reply services/masterdata/src/lib/core/services/masterdata-renderer.service.ts]: it was as you said... a code smell! trying to chain the javascript methods, basically generated the issue! something still seems wrong... i come back with feedback after some verification!

Comment (dageorgero) [reply services/masterdata/src/lib/core/services/masterdata-renderer.service.ts]: @Johanh-od changed the map with forEach as in my code case, forEach si more appropiate (I don't want to return a new array as with map).

**Review by bart-vandervliet — APPROVED** [review]

**Comment by bart-vandervliet** [comment]
What is the status of this PR? @dageorgero @Johanh-od?

**Comment by dageorgero** [comment]
@Johanh-od @gitout0fhere @jstorteboom i'm still waiting for 2 reviewers to approve this PR's. nothing i can do till than!

**Review by Johanh-od — APPROVED** [review]

**Pull request merged**